sensing/src/main/java/com/google/android/sensing/upload/SensingSynchronizer.kt

```diff
@@ -40,77 +44,96 @@ class SensingSynchronizer(
+                batchSyncUploadState as SyncUploadState.BatchState.InProgress,
+                it
+              )
+            emit(batchSyncUploadState)
+            // For terminal state of this batch we cancel his launch
+            if (batchSyncUploadState.isBatchTerminalState()) this@launch.cancel()
+          }
+        }
+
+      // await/join is needed to collect states completely
+      kotlin.runCatching { job.join() }.onFailure(Timber::w)
+
+      if (batchSyncUploadState is SyncUploadState.BatchState.Failed) awaitCancellation()
+
+      completedSyncRequests += uploadRequestList.size
+      uploadRequestList = uploadRequestFetcher.fetch()
+      totalSyncRequests += uploadRequestList.size
     }
+    emit(SyncUploadState.Completed(totalSyncRequests, completedSyncRequests))
   }
 
-  private fun calculateSyncUploadState(
-    lastSyncUploadState: SyncUploadState.InProgress,
+  private fun calculateSyncUploadStateInBatch(
+    lastSyncUploadState: SyncUploadState.BatchState.InProgress,
     uploadResult: UploadResult
-  ): SyncUploadState {
+  ): SyncUploadState.BatchState {
     /**
      * The last states cannot be terminal states like Completed, Failed, NoOp. Hence the only
-     * possible last states are [SyncUploadState.Started] & [SyncUploadState.InProgress].
+     * possible last states are [SyncUploadState.Started] & [SyncUploadState.BatchState].
      */
     with(lastSyncUploadState) {
       return when (uploadResult) {
         is UploadResult.Started -> {
-          SyncUploadState.InProgress(
-            totalRequests = totalRequests,
+          SyncUploadState.BatchState.InProgress(
+            totalBatchRequests = totalBatchRequests,
             completedRequests = completedRequests,
             currentTotalBytes = uploadResult.uploadRequest.fileSize,
             currentCompletedBytes = 0
           )
         }
         is UploadResult.Success ->
-          SyncUploadState.InProgress(
-            totalRequests = totalRequests,
+          SyncUploadState.BatchState.InProgress(
+            totalBatchRequests = totalBatchRequests,
             completedRequests = completedRequests,
             currentTotalBytes = currentTotalBytes,
             currentCompletedBytes = currentCompletedBytes + uploadResult.bytesUploaded
           )
         is UploadResult.Completed ->
-          if (completedRequests + 1 == totalRequests) SyncUploadState.Completed(totalRequests)
+          if (completedRequests + 1 == totalBatchRequests)
+            SyncUploadState.BatchState.Completed(totalBatchRequests)
           else
-            SyncUploadState.InProgress(
-              totalRequests = totalRequests,
+            SyncUploadState.BatchState.InProgress(
+              totalBatchRequests = totalBatchRequests,
               completedRequests = completedRequests + 1,
               currentTotalBytes = currentTotalBytes,
               currentCompletedBytes = currentCompletedBytes
```

Comment:
Should the `currentCompletedBytes` be something like `currentCompletedBytes = currentCompletedBytes + uploadResult.bytesUploaded` over here?

Reply:
That's the case when some chunk is uploaded successfully => UploadResult.Success

In this case UploadResult.Completed we merge all the uploads (not upload anything though)